source/mir/combinatorics/package.d

```diff
@@ -0,0 +1,1089 @@
+                result = result / i * m + result % i * m / i;
+            }
+            else
+            {
+                result = result * m / i;
+            }
+        }
+        return result;
+    }
+    else
+    {
+        import std.math: exp;
+        import std.mathspecial: logGamma;
+        real x = exp(logGamma(n+1)-logGamma(k+1)-logGamma(n-k+1));
+        return x;
+    }
+}
+
+///
+pure @safe unittest
+{
+    assert(binomial(5, 2) == 10);
+    assert(binomial(6, 4) == 15);
+    assert(binomial(3, 1) == 3);
+
+    import std.bigint;
+    assert(binomial!BigInt(1000, 10) == BigInt("263409560461970212832400"));
+}
+
+pure nothrow @safe @nogc unittest
+{
+    assert(binomial(5, 1) == 5);
+    assert(binomial(5, 0) == 1);
+    assert(binomial(1, 2) == 0);
+    assert(binomial(1, 0) == 1);
+    assert(binomial(1, 1) == 1);
+    assert(binomial(2, 1) == 2);
+    assert(binomial(2, 1) == 2);
+
+    // negative
+    assert(binomial!long(-5, 3) == -35);
+    assert(binomial!long(5, -3) == 0);
+}
+
+unittest
+{
+    import std.bigint;
+
+    // test larger numbers
+    assert(binomial(100, 10) == 17310309456440);
+    assert(binomial(999, 5) == 8209039793949);
+    assert(binomial(300, 10) == 1398320233241701770LU);
+    assert(binomial(300LU, 10LU) == 1398320233241701770LU);
+
+    // test overflow
+    assert(binomial(500, 10) == 0);
+
+    // all parameters as custom types
+    BigInt n = 1010, k = 9;
+    assert(binomial!BigInt(n, k) == BigInt("2908077120956865974260"));
+
+    // negative
+    assert(binomial!BigInt(-5, 3) == -35);
+    assert(binomial!BigInt(5, -3) == 0);
+
+    // floating point
+    import std.math: approxEqual;
+    assert(binomial!real(5.8, 3).approxEqual(17.632, 0.001));
+    assert(binomial!real(20.8, 3.2).approxEqual(1783.34, 0.01));
+    assert(binomial!real(-5.3, 4).approxEqual(84.2958375, 0.0001));
+    assert(binomial!real(-5.6, -3.4).approxEqual(0.00998410, 0.00001));
+    assert(binomial!real(-5.7, -7.6).approxEqual(20.1795, 0.0001));
+
+    // big floating point
+    assert(binomial!real(1.2e10, 2.3).approxEqual(5.66794e22, 0.0001));
+    // TODO: rounding errors seem to sum up here
+    //assert(binomial!real(2.245e15, 5.7).approxEqual(7.68342e84, 0.1));
+}
+
+/**
+Lazily computes all _permutations of `r` using $(WEB
+en.wikipedia.org/wiki/Heap%27s_algorithm, Heap's algorithm).
+
+While generating a new item is in `O(k)` (amortized `O(1)`),
+the number of permutations is `|r|!`.
+
+Returns:
+A forward range the elements of which are an $(XREF range,
+indexed) view into `r`.
+*/
+auto permutations(Range, Allocator)(Range r, auto ref Allocator alloc)
+if (isRandomAccessRange!Range && hasLength!Range)
+{
+    return Permutations!(Range, Allocator)(r, alloc);
+}
+
+/// ditto
+auto permutations(Range)(Range r)
+if (isRandomAccessRange!Range && hasLength!Range)
+{
+    return Permutations!Range(r);
+}
+
+private struct Permutations(Range, Allocator = void)
```

Comment:
Okay that looks nicer - thanks!

```
    this(Allocator)(Range r, ref Allocator alloc)
    {
        ...
    }

    void dispose(Allocator)(Allocator alloc)
    {
       ...
    }
```

But isn't it pretty inconvenient (and easy to forgot) for an user to dispose the structs internal state values manually?
AFAICT as soon as the structs gets destructed, they are not needed anymore. 

Moreover the other `structs` support `save`, which also needs allocation, thus saving the reference to the allocation is needed imho.